src/hotspot/cpu/s390/templateTable_s390.cpp

```diff
@@ -3116,10 +3223,13 @@ void TemplateTable::putfield_or_static(int byte_no, bool is_static, RewriteContr
   BIND(Done);
 
   // Check for volatile store.
-  Label notVolatile;
+  // only if flags register is non-volatile
+  NearLabel notVolatile;
 
-  __ testbit(Z_ARG4, ConstantPoolCacheEntry::is_volatile_shift);
-  __ z_brz(notVolatile);
+  if (!flags.is_volatile()) {
```

Comment:
This change is not legal. The property of the flag register has nothing to do with the bit in the content.

Reply:
Ahh. Wait. We just use the z_fence if the flags register was possibly killed?

Reply:
Yes, to keep it alive across VM-calls, we used non-volatile register here. 

We were short on registers and had to combine `obj and off` So that we free one register for `flags`. There exist comment for this, do you want me to add all of this logic explanation there ? 


```
 /*
   * In the static case, we can calculate the final field address easily.
   * Do so to occupy only one non-volatile register
   * ---------------------
   * In the non-static case, we preset fieldAddr with the field offset.
   * The object address is available only later. It is popped from stack.
   * see pop_and_check_object(obj);
   */
  if (is_static) {
    __ z_agr(fieldAddr, obj);
  }
```

Reply:
Ok, got it. Thanks for explaining. Please assert that the flag register is non-volatile instead.

Reply:
Done, thanks for the suggestion. Please review the latest changes.